File: Wire-iOS/Sources/UserInterface/ConversationList/ListContent/ConnectRequestsCell.m
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 


#import "ConnectRequestsCell.h"

@import PureLayout;

#import "ConversationListItemView.h"
#import "UIColor+WAZExtensions.h"

#import "WireSyncEngine+iOS.h"
#import "Constants.h"
#import "Wire-Swift.h"


@interface ConnectRequestsCell () <ZMConversationListObserver>

@property (nonatomic, strong) ConversationListItemView *itemView;
@property (nonatomic, assign) BOOL hasCreatedInitialConstraints;
@property (nonatomic, assign) NSUInteger currentConnectionRequestsCount;
@property (nonatomic) id conversationListObserverToken;

@end



@implementation ConnectRequestsCell

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        [self setupConnectRequestsCell];
    }
    return self;
}

- (void)setupConnectRequestsCell
{
    self.clipsToBounds = YES;
    self.itemView = [[ConversationListItemView alloc] initForAutoLayout];
    [self addSubview:self.itemView];
    [self updateAppearance];
    self.conversationListObserverToken = [ConversationListChangeInfo addObserver:self
                                                                         forList:[ZMConversationList pendingConnectionConversationsInUserSession:[ZMUserSession sharedSession]]
                                                                     userSession:[ZMUserSession sharedSession]];
    
    [self setNeedsUpdateConstraints];
}

- (void)updateConstraints
{
    if (! self.hasCreatedInitialConstraints) {
        self.hasCreatedInitialConstraints = YES;
        [self.itemView autoPinEdgesToSuperviewEdgesWithInsets:UIEdgeInsetsZero];
    }
    [super updateConstraints];
}

- (void)setSelected:(BOOL)selected
{
    [super setSelected:selected];
    if (IS_IPAD_FULLSCREEN) {
        self.itemView.selected  = self.selected || self.highlighted;
    }
}

- (void)setHighlighted:(BOOL)highlighted
{
    [super setHighlighted:highlighted];
    if (IS_IPAD_FULLSCREEN) {
        self.itemView.selected  = self.selected || self.highlighted;
    } else {
        self.itemView.selected  = self.highlighted;
    }
}

- (void)updateAppearance
{
    NSArray<ZMConversation *> *connectionRequests = [ZMConversationList pendingConnectionConversationsInUserSession:[ZMUserSession sharedSession]];
    
    NSUInteger newCount = connectionRequests.count;
    
    if (newCount != self.currentConnectionRequestsCount) {
        NSArray<ZMUser *> *connectionUsers = [connectionRequests mapWithBlock:^ZMUser *(ZMConversation *conversation) {
            return conversation.connection.to;
        }];
        
        self.currentConnectionRequestsCount = newCount;
        NSString *title = [NSString stringWithFormat:NSLocalizedString(@"list.connect_request.people_waiting", @""), newCount];
        [self.itemView configureWith:[[NSAttributedString alloc] initWithString:title]
                            subtitle:[[NSAttributedString alloc] init]
                               users:connectionUsers];
    }
}

- (void)conversationListDidChange:(ConversationListChangeInfo *)change
{
    [self updateAppearance];
}

@end

File: Wire-iOS/Sources/UserInterface/ConversationList/ListContent/ConnectRequestsCell.swift
// 
// Wire
// Copyright (C) 2019 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//

import WireSyncEngine

protocol SectionListCellType: AnyObject {
    var sectionName: String? { get set }
    var cellIdentifier: String? { get set }
}

extension SectionListCellType {
    var identifier: String {
        let prefix: String

        if let sectionName = sectionName {
            prefix = "\(sectionName) - "
        } else {
            prefix = ""
        }

        if let cellIdentifier = cellIdentifier {
            return prefix + cellIdentifier
        } else {
            return prefix
        }
    }
}

final class ConnectRequestsCell: UICollectionViewCell, SectionListCellType {
    var sectionName: String?
    var cellIdentifier: String?

    let itemView = ConversationListItemView()

    private var hasCreatedInitialConstraints = false
    private var currentConnectionRequestsCount: Int = 0
    private var conversationListObserverToken: Any?

    override init(frame: CGRect) {
        super.init(frame: frame)
        setupConnectRequestsCell()
    }

    @available(*, unavailable)
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    private func setupConnectRequestsCell() {
        clipsToBounds = true
        addSubview(itemView)
        updateAppearance()

        if let userSession = ZMUserSession.shared() {
            conversationListObserverToken = ConversationListChangeInfo.add(observer: self, for: ZMConversationList.pendingConnectionConversations(inUserSession: userSession), userSession: userSession)
        }

        setNeedsUpdateConstraints()
    }

    override var accessibilityIdentifier: String? {
        get {
            return identifier
        }
        set {
            // no op
        }
    }

    override func updateConstraints() {
        if !hasCreatedInitialConstraints {
            hasCreatedInitialConstraints = true
            itemView.translatesAutoresizingMaskIntoConstraints = false
            itemView.fitInSuperview()
        }
        super.updateConstraints()
    }

    private func updateItemViewSelected() {
        itemView.selected = isSelected || isHighlighted
    }

    override var isSelected: Bool {
        didSet {
            if isIPadRegular() {
                updateItemViewSelected()
            }
        }
    }

    override var isHighlighted: Bool {
        didSet {
            if isIPadRegular() {
                updateItemViewSelected()
            } else {
                itemView.selected = isHighlighted
            }
        }
    }

    private
    func updateAppearance() {
        guard let userSession = ZMUserSession.shared() else { return }

        let connectionRequests = ZMConversationList.pendingConnectionConversations(inUserSession: userSession)

        let newCount: Int = connectionRequests.count

        if newCount != currentConnectionRequestsCount {
            let connectionUsers = connectionRequests.map { conversation in
                if let conversation = conversation as? ZMConversation {
                    return conversation.connection?.to
                } else {
                    return nil
                }
            }

            if let users = connectionUsers as? [ZMUser] {
                currentConnectionRequestsCount = newCount
                let title = String(format: NSLocalizedString("list.connect_request.people_waiting", comment: ""), newCount)
                itemView.configure(with: NSAttributedString(string: title), subtitle: NSAttributedString(), users: users)
            }
        }
    }

}

extension ConnectRequestsCell: ZMConversationListObserver {
    func conversationListDidChange(_ changeInfo: ConversationListChangeInfo) {
        updateAppearance()
    }
}
